### ISSUE TITLE
Scaling Problems above ~15 workflows

### ISSUE THREAD

**Issue by cjllanwarne** [broadinstitute/cromwell#1065]
People are reporting significant slowdown when running more than about 15 workflows concurrently

**Comment by geoffjentry** [comment]
Perhaps a bit of extra detail?

**Comment by cjllanwarne** [comment]
Quoth Dave: Green Team launched 50 workflows and could initially query the API despite some slowness. After they’ve been running for 45 mins or so, hitting the API is only intermittently successful:

```https://cromwell.gotc-staging.broadinstitute.org/api/workflows/v1/query
Ooops! The server was not able to produce a timely response to your request.
Please try again in a short while!```

```Unexpected error while awaiting Cromwell Workflow completion: Error hitting REST API: https://cromwell.gotc-staging.broadinstitute.org/api/workflows/v1/5296889b-8b88-41db-a5fa-d1071ac22a... => Unexpected response code: 502```

Just trying to get to the swagger page takes a couple of minutes to load or fails to load altogether.

This is highly variable - about 2 hours in we still have 50 workflows running and API queries are coming back very quickly.

In production using Cromwell 0.19, GotC routinely runs ~200-500 workflows simultaneously.

**Comment by kcibul** [comment]
@dshiga is this how you ran the 50 before in testing?  Are they 50 simultaneous submissions?

**Comment by dshiga** [comment]
Yes, we ran the 50 the same way for a previous release and did not notice any slowness. I'm not sure we checked on it as much during the first couple of hours last time though. About 3 hours in now, it seems back to normal, no slowness.

The submissions are as close to simultaneous as we can get, meaning as fast as Cromwell will accept submissions from our script. Cromwell doesn't respond right away for each submission, so it takes about 5 minutes to submit 50.

**Comment by geoffjentry** [comment]
Had a brief chat w/ @dshiga - I was wondering if all slowness was directly tied to reading from metadata or general slowness and it was the latter (not that it couldn't be *caused* by reading from MD, but it's across the board slowness)

Three thoughts:
- We were talking yesterday about how someone not named me should try setting up typesafe monitor and use it to debug/profile/analyze. No time like the present!
- I can't find the ticket (@kcibul - do you know the right string to search for?) but we should look into the thread pools/dispatchers. This could be a case where something is bringing down the whole system but bulkheading would keep everything else responsive
- A while back we talked about streamlining submission such that WF submissions get a "Submitted" status but aren't necessarily immediately launched, and the system would pull them - allowing us to tune the rate at which we pull. Perhaps time to resurrect this one?

**Comment by mcovarr** [comment]
Stack traces show lots of threads on the default dispatcher backed up creating Google credentials like so:

```
"cromwell-system-akka.actor.default-dispatcher-49" #236 prio=5 os_prio=31 tid=0x00007faafe1ff800 nid=0x12903 waiting on condition [0x0000000133659000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007bb5f1698> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
	at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
	at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
	at com.google.api.client.auth.oauth2.Credential.refreshToken(Credential.java:486)
	at cromwell.filesystems.gcs.GoogleAuthMode$$anonfun$1.apply$mcZ$sp(GoogleAuthMode.scala:79)
	at cromwell.filesystems.gcs.GoogleAuthMode$$anonfun$1.apply(GoogleAuthMode.scala:79)
	at cromwell.filesystems.gcs.GoogleAuthMode$$anonfun$1.apply(GoogleAuthMode.scala:79)
	at scala.util.Try$.apply(Try.scala:192)
	at cromwell.filesystems.gcs.GoogleAuthMode$class.validateCredentials(GoogleAuthMode.scala:79)
	at cromwell.filesystems.gcs.ApplicationDefaultMode.validateCredentials(GoogleAuthMode.scala:138)
	at cromwell.filesystems.gcs.GoogleAuthMode$class.credential(GoogleAuthMode.scala:64)
	at cromwell.filesystems.gcs.ApplicationDefaultMode.credential(GoogleAuthMode.scala:138)
	at cromwell.filesystems.gcs.GoogleAuthMode$class.buildStorage(GoogleAuthMode.scala:95)
	at cromwell.filesystems.gcs.ApplicationDefaultMode.buildStorage(GoogleAuthMode.scala:138)
	at cromwell.backend.impl.jes.io.package$.buildFilesystem(package.scala:26)
	at cromwell.backend.impl.jes.JesWorkflowPaths.<init>(JesWorkflowPaths.scala:18)
	at cromwell.backend.impl.jes.JesInitializationActor.<init>(JesInitializationActor.scala:43)
	at cromwell.backend.impl.jes.JesInitializationActor$$anonfun$props$1.apply(JesInitializationActor.scala:22)
	at cromwell.backend.impl.jes.JesInitializationActor$$anonfun$props$1.apply(JesInitializationActor.scala:22)
```

The two problem actors here are JesAsyncBackendJobExecutionActor and JesInitializationActor.  JobPreparationActor also triggers a similar stack trace, but it's using the slow-actor-dispatcher so this doesn't adversely impact submissions.  Moving JesAsyncBackendJobExecutionActor and JesInitializationActor to slow-actor-dispatcher as well enabled me to submit up to 50 workflows immediately (I didn't try any more beyond this).

The auth / filesystem code was heavily reworked in #702 after 0.19_hotfix was branched.  There's a known issue with GCS filesystems not being reused #813 and it looks like there could be another issue with validating credentials excessively.
